File: docs/solutions/matlab/serial_stream_my_dc_motor.m
clc, clear all
%%

port = '/dev/ttyUSB0'; % port = 'COM12';
baudrate = 2e6;

% Initialize the SerialStream object
try
    serialStream.reset();
    fprintf("Resetting existing serialStream object.\n")
catch exception
    serialStream = SerialStream(port, baudrate);
    fprintf("Creating new serialStream object.\n")
end

% Starting the stream
serialStream.start()
while (serialStream.isBusy())
    pause(0.1);
end

% Accessing the data
try
    data = serialStream.getData();
catch exception
    fprintf("Data Stream not triggered.\n")
    return
end

% Save the data
file_name = 'data_my_dc_motor_00.mat';
save(file_name, 'data');

% Load the data
load(file_name)


%% Evaluate time

Ts = mean(diff(data.time));

figure(1)
plot(data.time(1:end-1), diff(data.time * 1e6)), grid on
title( sprintf(['Mean %0.0f mus, ', ...
                'Std. %0.0f mus, ', ...
                'Med. dT = %0.0f mus'], ...
                mean(diff(data.time * 1e6)), ...
                std(diff(data.time * 1e6)), ...
                median(diff(data.time * 1e6))) )
xlabel('Time (sec)'), ylabel('dTime (mus)')
xlim([0 data.time(end-1)])
ylim([0 1.2*max(diff(data.time * 1e6))])


%% Evaluate the data

% Defining the indices for the data columns
ind.counts    = 1;
ind.velocity  = 2;
ind.rotations = 3;
ind.voltage   = 4;
ind.velocity_setpoint = 5;
ind.velocity_target   = 6;

% Calculate smoothed acceleration
acceleration = [0; diff(data.values(:, ind.velocity))/Ts]; % prepend zero to match length
acceleration_smoothed = conv(acceleration, ones(1, 20)/20, 'same');

figure(2)
subplot(211)
plot(data.time, data.values(:, ind.voltage)), grid on
ylabel('Voltage (V)')
subplot(212)
plot(data.time, acceleration_smoothed), grid on
ylabel('Acceleration (RPS/sec)')
xlabel('Time (sec)')

figure(3)
subplot(311)
plot(data.time, data.values(:, ind.counts)), grid on
ylabel('Counts')
subplot(312)
plot(data.time, [data.values(:, ind.velocity_setpoint), ...
    data.values(:, ind.velocity_target), ...
    data.values(:, ind.velocity)]), grid on
legend('Setpoint', ...
    'Target', ...
    'Actual', ...
    'Location', 'best')
ylabel('Velocity (RPS)')
subplot(313)
plot(data.time, data.values(:, ind.rotations)), grid on
ylabel('Rotations')
xlabel('Time (sec)')

File: docs/solutions/python/serial_stream_my_dc_motor.py
# conda create --name pes-env python=3.11.4 numpy scipy matplotlib pyserial control ipykernel

import os
import time
import numpy as np
import matplotlib.pyplot as plt
import scipy as sp
from SerialStream import SerialStream


port = "/dev/ttyUSB0"  # "COM12"
baudrate = int(2e6)

# Initialize the SerialStream object
try:
    serialStream.reset()
    print("Resetting existing serialStream object.")
except Exception as e:
    serialStream = SerialStream(port, baudrate)
    print("Creating new serialStream object.")

# Starting the stream
serialStream.start()
while serialStream.is_busy_flag():
    time.sleep(0.1)

# Accessing the data
try:
    data = serialStream.get_data()
except Exception as e:
    print("Data Stream not triggered.")
    exit()

# Save the data
file_name = os.path.join("docs", "solutions" ,"python", "data_my_dc_motor_00.npz")
np.savez(file_name, **data)

# Load the data
loaded = np.load(file_name)
data = {"time": loaded["time"], "values": loaded["values"]}

# Evaluate time

Ts = np.mean(np.diff(data["time"]))

plt.figure(1)
plt.plot(data["time"][:-1], np.diff(data["time"]) * 1e6)
plt.grid(True)
plt.title(f"Mean {np.mean(np.diff(data['time']) * 1e6):.0f} mus, " f"Std. {np.std(np.diff(data['time']) * 1e6):.0f} mus, " f"Med. dT = {np.median(np.diff(data['time']) * 1e6):.0f} mus")
plt.xlabel("Time (sec)")
plt.ylabel("dTime (mus)")
plt.xlim([0, data["time"][-2]])
plt.ylim([0, 1.2 * np.max(np.diff(data["time"]) * 1e6)])

# Evaluate the data

# Defining the indices for the data columns
ind = {}
ind["counts"] = 0
ind["velocity"] = 1
ind["rotations"] = 2
ind["voltage"] = 3
ind["velocity_setpoint"] = 4
ind["velocity_target"] = 5

# Calculate smoothed acceleration
acceleration = np.insert(np.diff(data["values"][:, ind["velocity"]]) / Ts, 0, 0.0)  # prepend zero to match length
acceleration_smoothed = np.convolve(acceleration, np.ones(20) / 20, mode="same")

plt.figure(2)
plt.subplot(2, 1, 1)
plt.plot(data["time"], data["values"][:, ind["voltage"]])
plt.grid(True)
plt.ylabel("Voltage (V)")
plt.subplot(2, 1, 2)
plt.plot(data["time"], acceleration_smoothed)
plt.grid(True)
plt.ylabel("Acceleration (RPS/sec)")
plt.xlabel("Time (sec)")

plt.figure(3)
plt.subplot(3, 1, 1)
plt.plot(data["time"], data["values"][:, ind["counts"]])
plt.grid(True)
plt.ylabel("Counts")
plt.subplot(3, 1, 2)
plt.plot(data["time"], data["values"][:, ind["velocity_setpoint"]], label="Setpoint")
plt.plot(data["time"], data["values"][:, ind["velocity_target"]], label="Target")
plt.plot(data["time"], data["values"][:, ind["velocity"]], label="Actual")
plt.grid(True)
plt.ylabel("Velocity (RPS)")
plt.legend()
plt.subplot(3, 1, 3)
plt.plot(data["time"], data["values"][:, ind["rotations"]])
plt.grid(True)
plt.ylabel("Rotations")
plt.xlabel("Time (sec)")
plt.tight_layout()

# Show all plots
plt.show()
